### ISSUE TITLE
dive() breaks when using the new context API.

### ISSUE THREAD

**Issue by stellarhoof** [airbnb/enzyme#1647]
**Describe the bug**

The new React context elements (https://reactjs.org/docs/context.html) report their type as objects instead of functions (custom react components) or strings (native react components), which breaks `dive()`. I haven't tested whether other methods of `ShallowWrapper` break.

**To Reproduce**

https://codesandbox.io/s/zxj4898zxp?module=%2Fsrc%2FApp.test.js

**Expected behavior**

`shallow` and `dive` should implement actual render semantics when handling the new context API:
1. Automatically update current context whenever a context provider is encountered and remove the provider node from the tree.
2. Automatically pass the current context to childs of context consumer nodes and remove the consumer node from the tree.

_I'm sure the above are not the exact semantics, but at least the most intuitive._

**Comment by ljharb** [comment]
If you're using the latest version of `react-test-renderer` (you may need to `rm -rf node_modules && npm install`, or mess with your lockfile), is this still a problem?

**Comment by anthonator** [comment]
@ljharb I can confirm that I am seeing this error when using the latest version of `react-test-renderer`. I'm also using @KevinGrandon's `enzyme-react-adapter-future`. Not sure if that's relevant.

Also, just for reference, this is the error I'm seeing:

```
TypeError: ShallowWrapper::dive() can only be called on components
```

**Comment by anthonator** [comment]
Also, just for reference, this is what `debug` outputs on the `dive` just before receiving the above error:

```
<[object Object] value={{...}}>
  ...snip...
</[object Object]>
```

**Comment by anthonator** [comment]
@ljharb I believe @stellarhoof already provided a test case. Is there something more you need?

**Comment by anthonator** [comment]
@ljharb https://codesandbox.io/s/9l1rk4l20y?module=%2Fsrc%2FApp.test.js

Let me know if this is what you're looking for.

**Comment by joeldenning** [comment]
+1 I'm seeing the same problem.

**Comment by paddotk** [comment]
I just upgraded to Enzyme 3.4.1 but still getting the same error (`TypeError: ShallowWrapper::dive() can only be called on components`). Did the fix work for anyone else?

**Comment by ljharb** [comment]
@paddotk it's not been fixed yet.

**Comment by kseebrinegar** [comment]
Has this issue been fixed?

**Comment by ljharb** [comment]
nope, that's why it's still open

**Comment by kseebrinegar** [comment]
Sorry, about that.

**Comment by OzzieOrca** [comment]
Is there any way we can help with this? I tried digging into the code a little while ago and found it hard to follow. If I remember correctly, bypassing this check (and the error message it throws) https://github.com/airbnb/enzyme/blob/0bbdc988b149b8e75ff8cb530ab740b0459ce43d/packages/enzyme/src/ShallowWrapper.js#L1411 casued issues later on. I'm not really sure what needs to be done. I could try playing with it again.

I would love to move forward with an upgrade and this is the only thing holding us back as we've used dive a lot of places in our tests.

**Comment by ljharb** [comment]
@OzzieOrca the issue, I think, is a bit deeper than that. Although `shallow` takes a `context` option, that context is not filtered down to match the `contextTypes` of the component; `getChildContext()` is not taken into account, and the *new* child `context` is also not provided to the dived component and filtered down by `contextTypes`.

**Comment by OzzieOrca** [comment]
That's helpful, thanks. I might take a look at it again sometime soon unless someone more knowledgeable is able to jump on it.

**Comment by alexandrudanpop** [comment]
Is there a workaround for this issue? Getting this when testing redux connected components.

**Comment by ljharb** [comment]
@alexandrudanpop downgrade to redux < 6, iirc.

**Comment by OzzieOrca** [comment]
Unless I'm missing something redux is only at 4.0.1. https://github.com/reduxjs/redux/releases

**Comment by csvan** [comment]
I would love to help here if possible, can I get some pointers on where to start? Currently, this completely blocks us from moving to either `react-redux` 6.0 or `styled-components` 4.0, both which use the new Context API and completely break under `shallow`.

**Comment by ljharb** [comment]
@csvan the two pieces I'm waiting for are #1960 and then #1966; after those go in, I suspect it'll be relatively short work to get full createContext support.

**Comment by csvan** [comment]
@ljharb I apologise for being a nag, but what is the status on those PR:s being merged? Can we assist in any way?

**Comment by ljharb** [comment]
I've just landed #1960, and rebased #1966 (which needs some more work). I'm hopeful to make a release soon.
